lib/Doctrine/ORM/Internal/Hydration/AbstractHydrator.php

```diff
@@ -154,7 +155,7 @@ public function iterate($stmt, $resultSetMapping, array $hints = [])
      *
      * @return iterable<mixed>
      */
-    public function toIterable(Statement $stmt, ResultSetMapping $resultSetMapping, array $hints = []): iterable
+    public function toIterable(ResultStatement $stmt, ResultSetMapping $resultSetMapping, array $hints = []): iterable
```

Comment:
This looks a lot like a BC-break… under what circumstances do you observe the issue?

Reply:
I added a test.

Reply:
I doubt it's BC-break, because `Statement` extends `ResultStatement`. So it's the opposite of more restricting interface.

Reply:
It's a BC-break for extending classes, that now have to also relax their interface , and handle more types than they already do.

Oh but I now notice that `AbstractHydrator` is inside the `Internal` namespace, so that must be OK anyway I suppose :slightly_smiling_face:

Reply:
And the method below this one is 

```public function hydrateAll($stmt, $resultSetMapping, array $hints = [])`

No typehint at all.

Reply:
Yes but that's not suspicious: the git log shows that it was written at a time where type declarations did not even exist. Since the whole namespace is internal, I think it would be ok for you to add type hints if you want to. Maybe it was never reconsidered.

Reply:
Actually it seems it can be nullable ?
`
Failed asserting that exception of type "TypeError" matches expected exception "Doctrine\ORM\Query\QueryException". Message was: "Mock_AbstractHydrator_836a32f6::hydrateAll(): Argument #1 ($stmt) must be of type Doctrine\DBAL\Driver\ResultStatement, null given, called in /home/bozhidar_hristov/projects/orm/lib/Doctrine/ORM/AbstractQuery.php on line 1082" at
`

Reply:
It's might be because of the test itself:

 `$this->connection->method('executeQuery')->with($this->anything(), []);`

Reply:
Also it seems some of Docblock typehints are wrong:

`     
* @return array
*/
public function hydrateAll
`

but then

`13) Doctrine\Tests\ORM\Hydration\SingleScalarHydratorTest::testHydrateSingleScalar with data set #0 ('result2', array(array('1')))
TypeError: Doctrine\ORM\Internal\Hydration\AbstractHydrator::hydrateAll(): Return value must be of type array, int returned`

Reply:
Added typehints everywhere, just not on `$stmt` param on `hydrateAll`, because of failing tests.